### ISSUE TITLE
Use of 'before' with reporter causes "Cannot read property 'append' of undefined" error

### ISSUE THREAD

**Issue by Tom-Hudson** [serenity-js/serenity-js#104]
When I use `before` instead of `beforeEach` in the spec below, and I have the reporter enabled, it throws an error.

**Spec**
```
import { Open } from 'serenity-js/lib/screenplay-protractor';
import { stage } from './../../helpers/stage-manager';

import { invalid, valid } from './../../data/users';
import {
    CheckIfThePasswordField,
    CheckIfTheToastMessage,
    CheckIfTheUsernameField,
    ClickTheLoginButton,
    EnterThePassword,
    EnterTheUsername,
} from './login';

const anErrorMessageForInvalidCredentials = 'Username or password incorrect';

describe('Logging in', function() {
    this.timeout(60000);

    // Change this to beforeEach to pass
    before(() => stage.theActorCalled('Tom').attemptsTo(
        Open.browserOn('/login'),
    ));

    describe('with invalid credentials', () => {
        // Change this to beforeEach to pass
        before(() => stage.theActorInTheSpotlight().attemptsTo(
            EnterTheUsername.of(invalid.username),
            EnterThePassword.of(invalid.password),
            ClickTheLoginButton(),
        ));

        it('shows an error message', () => stage.theActorInTheSpotlight().attemptsTo(
            CheckIfTheToastMessage.contains(anErrorMessageForInvalidCredentials),
        ));

        it('clears the username field', () => stage.theActorInTheSpotlight().attemptsTo(
            CheckIfTheUsernameField.isBlank(),
        ));

        it('clears the password field', () => stage.theActorInTheSpotlight().attemptsTo(
            CheckIfThePasswordField.isBlank(),
        ));
    });
});
```

**Error**
```
  √ Logging in with invalid credentials shows an error message: 114ms
[09:13:31] E/launcher - Cannot read property 'append' of undefined
[09:13:31] E/launcher - TypeError: Cannot read property 'append' of undefined
    at C:\Projects\Automation Learning\SerenityJs\node_modules\@serenity-js\core\src\reporting\rehearsal_report.ts:30:46
    at Array.reduce (native)
    at Function.RehearsalReport.from (C:\Projects\Automation Learning\SerenityJs\node_modules\@serenity-js\core\src\reporting\rehearsal_report.ts:23:23)
    at SerenityBDDReporter.persistReport (C:\Projects\Automation Learning\SerenityJs\node_modules\@serenity-js\core\src\reporting\serenity_bdd_reporter.ts:51:29)
    at SerenityBDDReporter.notifyOf (C:\Projects\Automation Learning\SerenityJs\node_modules\@serenity-js\core\src\reporting\serenity_bdd_reporter.ts:45:45)
    at C:\Projects\Automation Learning\SerenityJs\node_modules\@serenity-js\core\src\stage\stage_manager.ts:112:53
    at Array.forEach (native)
    at CrewMembersCommunicationChannel.notify (C:\Projects\Automation Learning\SerenityJs\node_modules\@serenity-js\core\src\stage\stage_manager.ts:112:24)
    at StageManager.notifyOf (C:\Projects\Automation Learning\SerenityJs\node_modules\@serenity-js\core\src\stage\stage_manager.ts:24:52)
    at Serenity.notify (C:\Projects\Automation Learning\SerenityJs\node_modules\@serenity-js\core\src\serenity.ts:19:28)
[09:13:31] E/launcher - Process exited with error code 199

> serenity-js-trial@1.0.0 report C:\Projects\Automation Learning\SerenityJs
> serenity run

info: Using Java at: C:\ProgramData\Oracle\Java\javapath\java.EXE
info: All done!
Some of the npm scripts have failed:
====================================================================================================
protractor | serenity-js-trial@1.0.0 protractor: `protractor ./protractor.conf.js` Exit status 199 |

npm ERR! Test failed.  See above for more details.
```

If I simple change `before` to `beforeEach`, and do nothing else, it passes:
```
  √ Logging in with invalid credentials shows an error message: 629ms
  √ Logging in with invalid credentials clears the username field: 586ms
  √ Logging in with invalid credentials clears the password field: 580ms

  3 passing (20s)

[09:16:15] I/local - Shutting down selenium standalone server.
[09:16:15] I/launcher - 0 instance(s) of WebDriver still running
[09:16:15] I/launcher - chrome #01 passed

> serenity-js-trial@1.0.0 report C:\Projects\Automation Learning\SerenityJs
> serenity run

info: Using Java at: C:\ProgramData\Oracle\Java\javapath\java.EXE
info: All done!
```

However, the report site generated misses the top level description in the test output:
![image](https://user-images.githubusercontent.com/26221061/31811466-966db670-b577-11e7-9cd3-d1c3e413e0bf.png)

You can re-create both problems yourself with the [mocha-seed](https://github.com/serenity-js/seed-mocha) - but I did update a few packages in my project to see if these issues had been fixed!

**Versions**
chai: 4.1.2
chai-as-promised: 7.1.1
mocha: 4.0.1
npm-failsafe: 0.2.1
protractor: 5.1.2
serenity-cli: 0.7.0
serenity-js: 1.9.2
ts-node: 3.0.6
tslint: 5.4.3
typescript: 2.3.4

**Comment by jan-molak** [comment]
Thanks! Would you be able to raise a PR with a failing test demonstrating that? Or maybe even attempt a fix? [Here are the tests](https://github.com/jan-molak/serenity-js/tree/6fda43b5992d19228f36bc25214e5797b4499d35/packages/mocha/spec) checking the Mocha adapter.

I'm in a process of migrating from the [old Mocha adapter implementation](https://github.com/jan-molak/serenity-js/tree/6fda43b5992d19228f36bc25214e5797b4499d35/packages/serenity-js/src/serenity-mocha) to the [new one](https://github.com/jan-molak/serenity-js/tree/6fda43b5992d19228f36bc25214e5797b4499d35/packages/mocha/src) that will work both stand-alone and with protractor, so it would be great to fix the new one.

Thanks!
Jan

**Comment by hudson-t** [comment]
@jan-molak - I have forked the mocha-seed and made the change to re-create the issue [here](https://github.com/hudson-t/seed-mocha).

Sadly I am very much at the 'looking into using Serenity/JS' stage at work so the time I can spend on it in hours is limited. However, if I have time this week I will try and help out with a fix from home in the evenings.

Also, I am not sure my issue with the 'top level description' being missing is correct. Is the 'feature level' description meant to be shown in the report section I have in my screenshot? Looking at the format of the mocha-seed, I think not. If you can confirm this I will edit my original post and strike-through that bit.

**Comment by jan-molak** [comment]
Thanks, I appreciate you offering to contribute your free time to improve this project!

The [spec file for the Mocha reporter](https://github.com/jan-molak/serenity-js/blob/6fda43b5992d19228f36bc25214e5797b4499d35/packages/mocha/spec/recognises_scenario_results.spec.ts#L122) and the [documentation](http://serenity-js.org/mocha/automation.html#name-of-the-feature) should provide an answer to your question. 

At the moment, the outer-most `describe` block is recognised as the name of the feature being tested (akin to cucumber's [`Feature: ...` keyword](https://github.com/cucumber/cucumber/wiki/Feature-Introduction)).

Regarding the `before` block - this might be a bit of a tricky one. If you have a look at the [mocha reporter](https://github.com/jan-molak/serenity-js/blob/6fda43b5992d19228f36bc25214e5797b4499d35/packages/mocha/src/reporter.ts#L5), it listens on `test` and `test end` events (check out [this stackoverflow thread](https://stackoverflow.com/questions/18660916/how-can-i-subscribe-to-mocha-suite-events) to learn more about the events).

Based on what you've observed, it looks like invoking the `before` function does not trigger the `test` event, as opposed to invoking the `beforeEach`. This makes sense from Mocha's point of view, but also means that because the [`SceneStarts` event](https://github.com/jan-molak/serenity-js/blob/6fda43b5992d19228f36bc25214e5797b4499d35/packages/mocha/src/reporter.ts#L9) is not emitted until the test starts, any events [emitted by the `Actor`](https://github.com/jan-molak/serenity-js/blob/6fda43b5992d19228f36bc25214e5797b4499d35/packages/core/src/screenplay/actor.ts#L110) before that will get ignored in the [`RehearsalReport`](https://github.com/jan-molak/serenity-js/blob/e72f3b0cd118ff8b86b46c2b5a6895dffd32a0ac/packages/core/src/reporting/rehearsal_report.ts#L23) when it is created by the [`SerenityBDDReporter`](https://github.com/jan-molak/serenity-js/blob/e72f3b0cd118ff8b86b46c2b5a6895dffd32a0ac/packages/core/src/reporting/serenity_bdd_reporter.ts#L48).

Now, this begs a question how should the Serenity/JS-specific events, emitted before the test "officially" starts, get reported? We might need to introduce additional domain events that will be emitted when the `before` is called (probably the same case with `after`), and instruct the `RehearsalReport` to cache and replay those for each test perhaps?

So, it looks like you've just jumped from 'looking into using Serenity/JS' into its deepest parts! 

Welcome to the community :-)

J

**Comment by Tom-Hudson** [comment]
@jan-molak - Thanks for the info. As soon as my build issues are resolved I will crack on with fixing this. However, I am a little unsure about the whole caching of the events within `before` blocks, and putting them into each test. 

_**For a moment, please read on and humour me as I have a point...**_

The scenario in which I (and possibly others) want to 'do things' within a `before` block, is to clean test data from the database. Following the [description of an ability](http://serenity-js.org/design/screenplay-pattern.html#ability), I have currently implemented this by granting my actor the ability to `Interact.withThe(Database)`, which is wrapped in a task to `Start.afresh()`. so I can use it in a spec like so:

```
before(() => Sue.attemptsTo(
    Start.afresh()
));
```

Now I think about this, it feels incorrect to grant my actor the ability to `Interact.withThe(Database)`. This is because my actor is a 'Super User' on the system under test, who in the real world would never have the ability to interact with the database directly (thus, breaking the concept of my actor's role).

Instead, I feel this would be the job of a crew member, possibly a '[Stagehand](https://en.wikipedia.org/wiki/Stagehand)'. And maybe the "Stagehand's" actions do not need to be reported on as it is behind the scenes?

So, if instead of 'Sue' doing the Database stuff in the `before` block, the Stagehand would:

```
before(() => TheStagehand.attemptsTo(
    Clear.theStage()
));
```

Should a concept of having 'Stagehands' be introduced so 'setup' actions can be wrapped in a consistent way, which doesn't break the concept of having actors and still enables clear abstraction in your tests? - _there is currently to no real guidance on these things in the documentation, maybe I am wrong about all this and it just needs clarifying?._

**_back to my point now..._**

As I mentioned, I think the stagehand's actions shouldn't be reported which would resolve the issue in my case. But I do understand there will be a desire to report actions in a `before` block. It just feels incorrect to 'falsely' report certain actions occurred in all tests when they actually didn't. I don't really have an alternative suggestion for those cases, just thought I should share my thoughts before implementing your suggestion. Let me know what you think.

**Comment by jan-molak** [comment]
I love the metaphor of the Stagehand :-)

My typical approach is to have actors responsible for different aspects of the scenario, so you could have:

```typescript
const 
    Steve = Actor.named('Steve the Stagehand').whoCan(InteractWithTheDatabase.at('...')),
    Sue = Actor.named('Sue').whoCan(BrowseTheWeb.using(protractor.browser));
```

Having different types of actors playing different roles is perfectly consistent with the screenplay metaphor so I'm not sure if we need an explicit `Stagehand` role, or is an implicit one sufficient?

Now, if you don't want the Stagehand actions to be reported, you could implement their interactions without the `toString()` method on them. For example:

```typescript
class DropDBTable implements Interaction {
  performAs(actor: UsesAbilities & AnswersQuestions) {
    InteractWithTheDatabase.as(actor).dropTable(....);
  }

  // no toString
}
```

**Back to the issue with `before`**
I don't see a clean way of reporting tasks that take place in the `before` block yet... It doesn't mean that it doesn't exist though, I just not sure what's the best approach here.

I'm not a big fan of the "caching idea" to be perfectly honest, so I guess we should either explicitly ignore anything that happens in the `before` block (so that it doesn't cause errors), or decide on a way of reporting it. Maybe ignoring them explicitly for now is a sensible first step forward?

**Comment by rakeshnambiar** [comment]
@Tom-Hudson @jan-molak I too facing a similar issue. All my tests are PASS, however, error while generating the report and build is failing.

```
nodeSetup_1  | [e2e] 11 scenarios (11 passed)
nodeSetup_1  | [e2e] 176 steps (176 passed)
nodeSetup_1  | [e2e] 7m51.786s
nodeSetup_1  | [e2e] [13:01:08] I/launcher - 0 instance(s) of WebDriver still running
nodeSetup_1  | [e2e] [13:01:08] I/launcher - firefox #01 passed
zalenium     | 14:01:08.432 [http://172.19.0.5:40000] INFO  d.z.e.z.p.DockerSeleniumRemoteProxy - Session a96edf66-c129-4348-b77a-4960cd5fc7a0 completed. Node should shutdown soon...
nodeSetup_1  | [failsafe] Script 'e2e' exited with code 0
nodeSetup_1  | [report]
nodeSetup_1  | [report] > serenity-protractor-cucumber-pageobjects@1.0.0 report /usr/src/script
nodeSetup_1  | [report] > serenity run
nodeSetup_1  | [report]
nodeSetup_1  | [report] TypeError: Cannot read property '0' of undefined
nodeSetup_1  | [report]     at checkJavaHome.then.catch.then.then.catch.then.std (/usr/src/script/node_modules/java-home/lib/java-home.js:92:27)
nodeSetup_1  | [report] info: Using Java at: /usr/lib/jvm/java-11-openjdk-amd64/bin/java
nodeSetup_1  | [report] error: WARNING: An illegal reflective access operation has occurred
nodeSetup_1  | [report] serenity run
nodeSetup_1  | [report]
nodeSetup_1  | [report] Aggregates the JSON reports generated by Serenity BDD and produces one in HTML
nodeSetup_1  | [report]
nodeSetup_1  | [report] Options:
nodeSetup_1  | [report]   --version          Show version number                               [boolean]
nodeSetup_1  | [report]   --log              log level
nodeSetup_1  | [report]                          [choices: "info", "verbose", "debug"] [default: "info"]
nodeSetup_1  | [report]   --cacheDir         An absolute or relative path to where the Serenity BDD CLI
nodeSetup_1  | [report]                      jar file should be stored
nodeSetup_1  | [report]                    [default: "/usr/src/script/node_modules/serenity-cli/.cache"]
nodeSetup_1  | [report]   --destination      Directory to contain the generated Serenity BDD report
nodeSetup_1  | [report]                                [default: "/usr/src/script/target/site/serenity"]
nodeSetup_1  | [report]   --features         Source directory containing the Cucumber.js feature files
nodeSetup_1  | [report]                                            [default: "/usr/src/script/features"]
nodeSetup_1  | [report]   --source           Source directory containing the Serenity BDD JSON output
nodeSetup_1  | [report]                      files     [default: "/usr/src/script/target/site/serenity"]
nodeSetup_1  | [report]   --issueTrackerUrl  Base URL for issue trackers other than JIRA
nodeSetup_1  | [report]   --jiraProject      JIRA project identifier
nodeSetup_1  | [report]   --jiraUrl          Base URL of your JIRA server
nodeSetup_1  | [report]   --project          Project name to appear in the Serenity reports (defaults to
nodeSetup_1  | [report]                      the project directory name)
nodeSetup_1  | [report]   -h, --help         Show help                                         [boolean]
nodeSetup_1  | [report]
nodeSetup_1  | [report] Error: WARNING: An illegal reflective access operation has occurred
nodeSetup_1  | [report]     at /usr/src/script/node_modules/serenity-cli/lib/actions/logging.js:9:35
nodeSetup_1  | [report]     at process._tickCallback (internal/process/next_tick.js:68:7)
nodeSetup_1  | [report] npm ERR! code ELIFECYCLE
nodeSetup_1  | [report] npm ERR! errno 1
nodeSetup_1  | [report] npm ERR! serenity-protractor-cucumber-pageobjects@1.0.0 report: `serenity run`
nodeSetup_1  | [report] npm ERR! Exit status 1
nodeSetup_1  | [report] npm ERR!
nodeSetup_1  | [report] npm ERR! Failed at the serenity-protractor-cucumber-pageobjects@1.0.0 report script.
nodeSetup_1  | [report] npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
nodeSetup_1  | [report]
nodeSetup_1  | [report] npm ERR! A complete log of this run can be found in:
nodeSetup_1  | [report] npm ERR!     /root/.npm/_logs/2019-03-12T13_01_12_686Z-debug.log
nodeSetup_1  | [failsafe] Script 'report' exited with code 1
nodeSetup_1  | npm ERR! Test failed.  See above for more details.
```

**Comment by jan-molak** [comment]
This was a synchronisation error in `SerenityBDDReporter`, addressed in [Serenity/JS 2.0](https://serenity-js.org/modules/serenity-bdd).

Serenity/JS 2.0 does not have a Mocha-specific adapter (yet!), but in the meantime, you can use it with Jasmine - see the [template repo](https://github.com/serenity-js/serenity-js-jasmine-protractor-template),
